### PR TITLE
Readability edits: Writing good specifications section

### PR DESCRIPTION
**Pull request by alice**
***
<a href="https://pr-preview.s3.amazonaws.com/alice/design-principles/pull/227.html" title="Last updated on Nov 3, 2020, 8:36 AM UTC (9205acb)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/227/d0ad23a...alice:9205acb.html" title="Last updated on Nov 3, 2020, 8:36 AM UTC (9205acb)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -2055,86 +2055,98 @@ for the APIs that they design.
...
+- Clearly specify the inputs and outputs,
+name the algorithm and the variables it uses well,
+and explicitly note the points in the algorithm where the algorithm may
+return a result or error.
+- As much as possible, avoid writing algorithms which have side effects.
+
+Summarize the purpose of the algorithm before going into detail,
+so that readers can decide whether to read the steps or skip over them.
+For example take the following steps, which ensure that there is at most one pending X
+callback per toplevel browsing context".
```
Comment (plinss) [inline index.bs:2131]: Missing opening quote

File: index.bs [hunk 2]
```diff
@@ -2055,86 +2055,98 @@ for the APIs that they design.
...
+
+If a specification isn't specific enough,
+implementers might make different choices
+which force authors to write extra code to handle the differences.
+
+Implementers shouldn't need to check details of other implementations
+to avoid this situation.
+Instead, the specification should be complete and clear enough
+on its own.
+
```
Comment (plinss) [inline index.bs:2101]: Possibly add a note to implementers here to file issues on a spec when the necessary level or clarity isn't present

File: index.bs [hunk 3]
```diff
@@ -2055,86 +2055,98 @@ for the APIs that they design.
...
+Issue: I don't understand the paragraph below.
 
 When specifying in an algorithmic way,
 it is generally good to match the algorithms that implementations would use,
 even if this is somewhat less convenient for specification authors,
-as long as doing so would not significantly increase the specification's complexity.
-For example, it is better to specify CSS selector matching
+as long as doing so wouldn't significantly increase the specification's complexity.
+For example, it's better to specify CSS selector matching
```
Comment (alice) [inline index.bs:2149]: https://drafts.csswg.org/selectors-4/#match-against-element
goes right-to-left.

Selector matching is done in two ways - during style, you start with an element, and go from right to left, and during querySelector, when you would go left to right...

File: index.bs (hunk 1 above)
Comment (alice) [inline index.bs:2131]:
```suggestion
callback per toplevel browsing context.
```

Comment (cynthia) [reply index.bs]: Also, maybe "top-level" instead of toplevel.

File: index.bs [hunk 4]
```diff
@@ -2055,86 +2055,98 @@ for the APIs that they design.
...
+- Clearly specify the inputs and outputs,
+name the algorithm and the variables it uses well,
+and explicitly note the points in the algorithm where the algorithm may
+return a result or error.
+- As much as possible, avoid writing algorithms which have side effects.
+
+Summarize the purpose of the algorithm before going into detail,
+so that readers can decide whether to read the steps or skip over them.
+For example take the following steps, which ensure that there is at most one pending X
+callback per toplevel browsing context.
```
Comment (cynthia) [inline index.bs:2131]:
```suggestion
callback per top-level browsing context.
```

File: index.bs [hunk 5]
```diff
@@ -2055,86 +2055,98 @@ for the APIs that they design.
 <h3 id="requirements-on-authors-and-implementers">Identify the audience of each requirement in your specification</h3>
 <!-- was "Conformance for Documents and Implementations" in the HTML Design Principles -->
 
-The web platform is,
-unusually,
+Document both how authors should write good code using your API,
+and how implementers of your API should handle poorly-written code.
+
+The web platform is, unusually,
```
Comment (torgo) [inline index.bs:2061]:
```suggestion
unusually when compared to other platforms, the web is
```

File: index.bs [hunk 6]
```diff
@@ -2145,27 +2157,15 @@ It also helps to keep the costs of specifying features aligned with implementing
...
-callback per toplevel browsing context")
-so that readers can quickly decide whether they need to read the steps in detail.
+See also:
+* <a href="https://infra.spec.whatwg.org/#algorithms">some useful definitions
+and terminology</a> from [[!INFRA]]
 
 <h4 id="use-flags-for-states">Use explicit flags for state</h4>
 
-Describing state using explicit flags makes behavior clearer.
+Instead of describing state in prose,
```
Comment (cynthia) [inline index.bs:2166]:
```suggestion
Instead of describing state with words,
```

File: index.bs [hunk 7]
```diff
@@ -2055,86 +2055,98 @@ for the APIs that they design.
...
+
+A plain sequence of steps is not always the best way to write an algorithm.
+For example, it might make sense to define
+or re-use a formal syntax or grammar
+to avoid repetition,
+or define specific states to be used in a state machine.
+When using extra constructs like these,
+the earlier advice still applies.
+
+Issue: I don't understand the paragraph below.
```
Comment (cynthia) [inline index.bs:2143]: I believe this comment is suggesting the pseudocode in the spec should not be simplified to a point that it is unrealistic for directly translation to implementation code. (e.g. fib() in recursive vs imperative, given a non-LISP setup.)

At least that's how I understood it.

File: index.bs [hunk 8]
```diff
@@ -2055,86 +2055,98 @@ for the APIs that they design.
...
+Summarize the purpose of the algorithm before going into detail,
+so that readers can decide whether to read the steps or skip over them.
+For example take the following steps, which ensure that there is at most one pending X
+callback per top-level browsing context.
+
+Issue: The paragraph below may not be comprehensive enough.
+
+A plain sequence of steps is not always the best way to write an algorithm.
+For example, it might make sense to define
+or re-use a formal syntax or grammar
```
Comment (cynthia) [inline index.bs:2137]: Might be good to link to an example or definition of "formal syntax or grammar"

Comment (torgo) [reply index.bs:2140]: https://en.wikipedia.org/wiki/Backus–Naur_form ?

File: index.bs [hunk 9]
```diff
@@ -2055,117 +2055,124 @@ for the APIs that they design.
...
+return a result or error.
+- As much as possible, avoid writing algorithms which have side effects.
+
+Summarize the purpose of the algorithm before going into detail,
+so that readers can decide whether to read the steps or skip over them.
+For example take the following steps, which ensure that there is at most one pending X
+callback per top-level browsing context.
+
+Issue: The paragraph below may not be comprehensive enough.
+
```
Comment (alice) [inline index.bs:2137]:
```suggestion
```

File: index.bs (hunk 8 above)
Comment (torgo) [reply index.bs:2138]: ... or https://tools.ietf.org/html/rfc5234 ?

Comment (torgo) [reply index.bs:2138]: https://www.w3.org/TR/CSS21/grammar.html#grammar ?

Comment (alice) [reply index.bs:2138]: Let's leave this as a separate issue.

**Review by torgo — APPROVED** [review]

**Review by cynthia — APPROVED** [review]

**Pull request merged**